elastic/integrations
# Add condition to leverage kubernetes provider for controllermanager and scheduler data streams

## What does this PR do?

This PR adds a condition in controller manager and scheduler data streams so as to leverage kubernetes dynamic provider.
Only agent running on k8s node where `Kube-controller-manager` and `Kube-scheduler` are also running(control plane) will fetch metrics.

## Checklist


## How to test this PR locally

1. Build the kubernetes package `elastic-package build` inside `integrations/packages/kubernetes` folder
2. Bring up elastic stack `elastic-package  stack up --version=7.15.0-SNAPSHOT -v -d`
3. Set up multi-node k8s cluster with kind (https://kind.sigs.k8s.io/docs/user/quick-start/#multinode-clusters)
4.  Connect kind nodes to elastic stack network [docker-connect.sh.zip](https://github.com/elastic/integrations/files/6807091/docker-connect.sh.zip)
5. In Kibana ui create a new agent policy with `kube-controller-manager` and `kube-scheduler` integration. Activate `Collect Kubernetes metrics from Kubernetes Scheduler` and `Collect metrics from Kubernetes controller-manager`.  Save the integration to the policy and copy the enrolment token of the new policy.
6. Update [elastic-agent-managed](https://github.com/elastic/beats/tree/master/deploy/kubernetes/elastic-agent-managed) daemonset variables (`FLEET_URL`: `"http://fleet-server:8220"`, `FLEET_ENROLLMENT_TOKEN`: `token from step 5` and set image to `docker.elastic.co/beats/elastic-agent:7.15.0-SNAPSHOT` . Apply the manifests `kubectl apply -f .` 
7.  After agents are successfully enrolled to fleet and assigned the policy check in discovery that `controller-manager` and `scheduler` metrics (for example `metricset=controllermanager`) come only from `agent.name==kind-control-plane` which is the agent running on the node that `controller-manager` is running.

In a scenario where `controller manager`  does not have `component=kube-controller-manager` label already set, the user can set a custom label like `my_label=controller` to the running pod and in step 5 can specify in advanced options of `kube-controller-manager` integration the label key and value. After applying the policy the agent running in `kind-control-plane` will start fetching metrics from the pod.
## Screenshots
<img width="1568" alt="agents" src="https://user-images.githubusercontent.com/26270880/125922234-9da68b50-185a-488e-8ca1-cc81e60d9603.png">
<img width="1779" alt="discover-controllermanager" src="https://user-images.githubusercontent.com/26270880/125922730-13ee11fc-83c3-4693-8409-309e3febc00f.png">

<img width="1669" alt="labels" src="https://user-images.githubusercontent.com/26270880/129060645-9d1a093b-4659-44f3-81ac-d89d209b3553.png">
<img width="1373" alt="custom_label" src="https://user-images.githubusercontent.com/26270880/129060680-1d188198-e777-422a-87c5-51bcde53e8ba.png">
